typescript/infra/src/core/govern.ts

```diff
@@ -77,25 +94,70 @@ export class AbacusCoreGovernor<Chain extends ChainName> {
     });
   }
 
-  async handleValidatorViolation(violation: ValidatorViolation) {
-    const validatorManager = violation.data.validatorManager;
-    switch (violation.data.type) {
-      case ValidatorViolationType.EnrollValidator: {
-        const call = await validatorManager.populateTransaction.enrollValidator(
-          violation.expected,
-        );
-        this.pushCall(violation.chain as Chain, call as types.CallData);
+  // pushes calls which reconcile actual and expected sets on chain
+  protected async pushSetReconcilationCalls<T>(reconcile: {
+    chain: ChainName;
+    actual: Set<T>;
+    expected: Set<T>;
+    add: (elem: T) => Promise<PopulatedTransaction>;
+    remove: (elem: T) => Promise<PopulatedTransaction>;
+  }) {
+    let txs: PopulatedTransaction[] = [];
+    // add expected - actual elements
+    utils
+      .difference(reconcile.expected, reconcile.actual)
+      .forEach(async (item) => txs.push(await reconcile.add(item)));
```

Comment:
I don't love that we don't await thse for-eached asyncs

Reply:
forEach isnt async iirc